_posts/2020-06-05-life-locks.md

```diff
@@ -0,0 +1,155 @@
+
+1. [s3:object-lock-mode](https://docs.aws.amazon.com/AmazonS3/latest/dev/list_amazons3.html#amazons3-policy-keys) is used to enforce which locking mode is used; and
+2. [s3:object-lock-remaining-retention-days](https://docs.aws.amazon.com/AmazonS3/latest/dev/list_amazons3.html#amazons3-policy-keys) is used to enforce the required lock duration.
+
+#### s3:object-lock-mode
+
+By using `s3:object-lock-mode` you can enforce which [S3 Actions](https://docs.aws.amazon.com/IAM/latest/UserGuide/list_amazons3.html#amazons3-actions-as-permissions) require a lock mode value (the first stanza), as well as the requirement that it has a value of **COMPLIANCE** (second stanza).
+
+```yaml
+- Sid: 'RequireObjectLockMode'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-mode': 'true'
+- Sid: 'RequireObjectLockModeToBeCompliance'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'StringNotEqualsIgnoreCase':
+        's3:object-lock-mode': 'COMPLIANCE'
+```
+
+#### s3:object-lock-remaining-retention-days
+
+Next you enforce how long uploaded files are protected by using the `s3:object-lock-remaining-retention-days` condition key. In the example below, you require that a retention period be specified, and that it must be 7 days. It is also possible to use this condition key to enforce a day range; I'll leave that as an exercise for you dear reader :wink:
+
+```yaml
+- Sid: 'RequireRetentionDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      'Null':
+        's3:object-lock-remaining-retention-days': 'true'
+- Sid: 'RequireRetentionToBeThreeDays'
+  Effect: Deny
+  Principal: '*'
+  Action:
+    - 's3:PutObject'
+    - 's3:PutObjectRetention'
+  Resource:
+    - !Sub '${Bucket.Arn}/*'
+  Condition:
+      NumericNotEquals:
+        's3:object-lock-remaining-retention-days': '7'
+```
+
+### Unspoken Requirement
+
+The last and unspoken requirement can be found in this [complete CloudFormation template](/assets/2020-06-03-life-locks-example-template.yml).
+
+## Definitions
+
+During your reading you realise that some of the terms in [Object Locking](https://docs.aws.amazon.com/AmazonS3/latest/dev/object-lock.html) are a bit vague, so you document them for the rest of the team:
+
+* **LegalHold**: view this as making an object 'read-only'. It is independent to the `LockingMode`; you can specify neither, either or both of Legal Hold and Locking Mode.
+* **LockingMode**: view this has how tamper proof the WORM model is, and has one of the following values:
```

Comment:
should this read "view this as" instead of "view this has"

Reply:
Yep, fixed.